src/controller/userController.ts

```diff
@@ -0,0 +1,50 @@
+import { registerUser } from "../services/registerUser";
+import { generateCrypto } from "../services/generateCrypto";
+import { sendConfirmationMail } from "../services/sendEmail";
+import type { RequestHandler } from "express";
+import { parseUserRegisterCredentials } from "../services/userCredentialsParser";
+import type { ZodError } from "zod";
+import { insertActivationCode } from "../services/databaseServices/insertActivationCode";
+import { findUserBy } from "../services/databaseServices/findUserBy";
+
+export const userRegisterController: RequestHandler = async (
+  req,
+  res,
+  _next
+) => {
+  const { username, password, email } = req.body;
+  const userCredentials = await parseUserRegisterCredentials({
+    username,
+    password,
+    email,
+  });
+  if (userCredentials[0]) {
```

Comment:
Awful API of `parseUserRegisterCredentials`...
it either returns validation errors or the input params (?)

Honestly, if you really want to avoid throwing the error I see it this way (introducing an object holding the parsing result):
```
const parseResult = parseUserRegisterCredentials(...);
if(parseResult.hasErrors()) {
  return res.status(403).json({ errors: validationResult.errors });
}

const { username, password, email } = parseResult.data;
```